### PR TITLE
[FRONT-158, FRONT-162] Stream domains (ENS + eth accounts)

### PR DESCRIPTION
**Pull request by juhah**
Stream domain selection revisited. This changed from the default `sandbox` to include user's connected eth accounts instead:

Screenshot 2020-11-20 at 15 12 56

The ENS accounts are still fetched but by default the user's username or the first connected account will be the default domain. If no accounts are connected, there is a link to add a new account.

Since the ENS API is connected to the mainnet, I couldn't test a real world case with the ENS domains. I used a hardcoded address to test it and separately verified that the query sends the addresses (= connected accounts) of the current user.

The domains show up in the edit & read-only page like this:

Screenshot 2020-11-20 at 15 12 32

Screenshot 2020-11-20 at 15 13 17

For legacy streams without a domain/path configuration, we only show the stream id:

Screenshot 2020-11-20 at 15 14 30
Screenshot 2020-11-20 at 15 14 56

When testing, remember to update your docker to the latest version to be able to create streams under your own ETH accounts!

### CODE REVIEW

File: app/src/userpages/components/StreamPage/New.jsx [hunk 1]
```diff
@@ -173,18 +253,90 @@ const UnstyledNew = (props) => {
...
-        value: DEFAULT_DOMAIN,
-        label: DEFAULT_DOMAIN,
-    },
-        // ... todo: get domains
-    /* {
-        value: undefined,
-        label: 'Add new domain',
-    } */]), [])
+    const loadDomains = useCallback(async () => {
+        const addresses = ethereumIdentitiesRef.current.map(({ json }) => json.address)
```
Comment (juhah) [inline app/src/userpages/components/StreamPage/New.jsx:264]: If you replace the addresses here with some that own domains, eg.

```js
// const addresses = ethereumIdentitiesRef.current.map(({ json }) => json.address)
const addresses = ['0xa729addefe1fa7bce87053ed55d55edddd13de60']
```

You can see some actual results.

File: app/src/userpages/components/StreamPage/New.jsx [hunk 2]
```diff
@@ -126,12 +207,11 @@ const getValidId = ({ domain, pathname }) => {
...
     }
 
-    if (!/\w$/.test(id)) {
+    if (!(/\w$/.test(id))) {
         throw new Error(I18n.t('userpages.streams.validation.invalidPathEndCharacter'))
     }
 
-    // this matches the backend validation for the full id
-    if (!/^((?:[\w-]+\.?)*\w)\/(?:[\w.-]+\/?)*\w$/.test(id)) {
+    if (/[^\w.\-/]/.test(id)) {
```
Comment (juhah) [inline app/src/userpages/components/StreamPage/New.jsx:214]: The previous rule caused a forever loop with a path containing spaces

File: app/src/shared/components/Spinner/index.jsx [hunk 3]
```diff
@@ -7,7 +7,7 @@ import { Translate } from 'react-redux-i18n'
 import styles from './spinner.pcss'
 
 export type SpinnerSize = 'small' | 'large'
-export type SpinnerColor = 'green' | 'white' | 'gray'
+export type SpinnerColor = 'green' | 'white' | 'gray' | 'blue'
```
Comment (juhah) [inline app/src/shared/components/Spinner/index.jsx:10]: New spinner type

Screenshot 2020-11-20 at 15 28 55

File: app/src/marketplace/containers/EditProductPage/DeployDataUnionModal.jsx [hunk 4]
```diff
@@ -72,15 +72,17 @@ export const DeployDialog = ({ product, api, updateAddress }: DeployDialogProps)
             throw new Error('no product!')
         }
 
-        const { id: joinPartStreamId } = await createJoinPartStream(productId)
+        const web3 = getWeb3()
+        const account = await web3.getDefaultAccount()
+        const { id: joinPartStreamId } = await createJoinPartStream(account, productId)
```
Comment (juhah) [inline app/src/marketplace/containers/EditProductPage/DeployDataUnionModal.jsx:77]: This works for now, join/part stream creation is going away and will any change a bit in #1109

File: app/src/marketplace/modules/dataUnion/services.js [hunk 5]
```diff
@@ -27,9 +27,9 @@ export const getStreamrEngineAddresses = (): Array<string> => {
     return addresses
 }
 
-export const createJoinPartStream = async (productId: ProductId): Promise<Stream> => {
+export const createJoinPartStream = async (account: Address, productId: ProductId): Promise<Stream> => {
     const newStream: NewStream = {
-        id: `sandbox/dataunions/${productId}/joinPartStream`,
+        id: `${account}/dataunions/${productId}/joinPartStream`,
```
Comment (juhah) [inline app/src/marketplace/modules/dataUnion/services.js:32]: Getting/setting permissions with `dataunions` in the path seems to throw an error in backend 🤔

Comment (juhah) [reply app/src/marketplace/modules/dataUnion/services.js:32]: https://linear.app/streamr/issue/BACK-100/cannot-get-stream-permission-with-dataunion-in-path

**Review by mondoreale — APPROVED** [review]
Looks good to me.

The new "Create stream" button… any chance there's a discussion about making stream creation a 2-step process? Current "New stream" page is just a place where you set up the id. Then you're taken to the edit page. Sure they're the same, pretty much. I'm not quite sure if I like seeing all them disabled fields. Just a thought.

Cheers!


PS. Was easier cause I've already gone through 1 version of it.

File: app/src/userpages/components/StreamPage/shared/FormGroup.jsx [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        justify-content: space-between;
+    }
+
+    ${Text} {
+        width: 100%;
+    }
+`
+
+export const StreamIdFormGroup = styled(FormGroup)`
+    & + ${FormGroup} {
```
[the remark this answers is not in the document]
Comment (juhah) [reply app/src/userpages/components/StreamPage/shared/FormGroup.jsx:75]: You would think but seems like it's not being covered by that rule.

File: app/cypress/integration/StreamPage.spec.js [hunk 7]
```diff
@@ -111,72 +111,83 @@ describe('New stream page', () => {
         cy.login()
         cy.visit('/core/streams/new')
         cy.location('pathname').should('eq', '/core/streams/new')
-
         cy.get('[data-test-hook=TOCPage]').find('h1').contains(/name your stream/i)
-        cy.get('[name=domain]').invoke('val').should('eq', 'sandbox')
+
+        // Wait for domains to load
+        // eslint-disable-next-line cypress/no-unnecessary-waiting
+        cy.wait(3000)
```
[the remark this answers is not in the document]
Comment (juhah) [reply app/cypress/integration/StreamPage.spec.js:118]: Yep, i tried it before and it seemed quite flaky... easier to wait for it this way. I still want to add couple of more tests so I will try again then 👍

**Comment by juhah** [comment]
I'll merge this to get things working against the current backend - let's revisit any tweaks later.

**Pull request merged**